### ISSUE TITLE
Assess range of array values during metrics computation

### ISSUE THREAD

**Issue by hermancollin** [ivadomed/ivadomed#868]
### Motivation for the feature

While training some microscopy models, I encountered dice scores superior to 1 during the testing phase. Investigating this further, I noticed that the dice score computation [here](https://github.com/ivadomed/ivadomed/blob/master/ivadomed/metrics.py#L91) used a "float" representation for one operand (MIN 0.0 - MAX 1.0) and an 8-bit integer value (stored as a float64) for the other operand (MIN 0.0 - MAX 255.0). @mariehbourget's hypothesis for this is that data read from the NiFTI files is converted during the training pipeline but not during the microscopy testing pipeline.
Another user previously encountered something similar in #468.

### Description of the feature

Besides fixing this by adding the right conversion during the testing phase, I would suggest adding some form of check for this because this kind of bug invalidates the eval metrics. Something like this:
```python
if im1.max() != im2.max():
    raise ValueError("Range mismatch: im1 and im2 must be on the same scale.")
```

**Comment by mariehbourget** [comment]
Good catch @hermancollin, thank you!
I ran a few short tests with MRI and microscopy data.

### MRI data
Example in training, where im1 is the prediction and im2 the GT from https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/training.py#L257
```
im1 max: 0.56155044
im2 max: 1
```
Example in testing, where im1 is the prediction and im2 the GT from https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/testing.py#L78
```
im1 max: 0.9998077750205994
im2 max: 1.0
```
Example in evaluation, where im1 is the GT and im2 the prediction from https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/evaluation.py#L430 **Note: the order of pred and GT are inverted in evaluation, it doesn't change the result but could create confusion when debugging**
```
im1 max: 1.0
im2 max: 0.9998078132048249
```

### Microscopy data
Example in training, where im1 is the prediction and im2 the GT:
```
im1 max: 0.5082986
im2 max: 1
```
Example in testing, where im1 is the prediction and im2 the GT:
```
im1 max: 1.0
im2 max: 255.0
```
Example in evaluation, where im1 is the GT and im2 the prediction:
```
im1 max: 255.0
im2 max: 1.0
```

So as suspected, the range for the GT is 0-1 in training but not in testing/eval. We will have to check the loader to see where the conversion happens for training. Then see if we should convert the levels when writing the NifTi files from PNG in the loader, or when feeding the GT to testing/eval.

For the check, we should verify that the range is between 0 and 1, as commented in https://github.com/ivadomed/ivadomed/issues/468#issuecomment-705262123

**Comment by mariehbourget** [comment]
**Training**: The GT values are read from the batch of the torch DataLoader and they seem to be normalized between 0 and 1.
https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/training.py#L255 

**Testing**: The GT values are read "as-is" from the NifTI file. https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/testing.py#L243

**Evaluation**: The GT values are also read "as-is" from the NifTI file. https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/evaluation.py#L81

Conclusion: This is not a "microscopy" issue, any GT with values over 1 will present the same problem as detected by @hermancollin, also documented in #468. So the idea to normalize the value when converting the microscopy images to NifTI is not good.

I think we should check the range and normalize the values of the GT in range 0-1 before computing the metrics in both testing and evaluation, as metrics other than dice_score() are also affected. @charleygros, I would like to have you input on that. Thanks!

**Comment by charleygros** [comment]
@mariehbourget & @hermancollin : you guys did a great job at digging this out and investigating things.

I agree, I think you well identified the root of the problem.

> I think we should check the range and normalize the values of the GT in range 0-1 before computing the metrics in both testing and evaluation.

It could fix @hermancollin 's bug. But let's be careful: we want to be robust to non-binary tasks, eg vertebral labelling: C2 with label 2, C3 with label 3 etc.
An option would be to bring this back to a binary (as we do for multiclass segmentation): eg 7 vertebral labels: GT would have 7 channels, all binary, channel1 for first vertebral label, channel2 for second vertebral label. Could be done in the loader?
Cons: (i) can quickly be heavy memory-wise, (ii) what if in the future we deal with tasks encompassing 100 classes.

Would be good to discuss this matter during next meeting. But please y'all: write down and share your thoughts down here. Cheers

**Comment by mariehbourget** [comment]
Thanks @charleygros for the feedback! I was not aware of the classification task you mentioned and thought the range was always "squeezed" between 0 and 1 as a float instead of binarized to 0 or 1 as an int.

If I understand correctly, there are 3 situation that we have to account for:

**1) Segmentation with "hard" labels** where GT values are binarized to 0 or 1 (int) after data augmentation with hardcoded threshold of 0.5 here:
https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/loader/mri2d_segmentation_dataset.py#L218
- Are labels for MRI segmentation always of value 1 (hence the threshold of 0.5)?
- For Microscopy segmentation, it works for training with hard labels because of GT values of 255 > 0.5, but it's not ideal since the 0.5 seems to assume a range of 0-1. Moreover it fails in evaluation because of the range 0-255, instead of 0-1.

**2) Segmentation with "soft" labels** where GT values are left "as-is" in float after data augmentation.

**3) Classification** where each GT value (ex: 1, 2, 3) represent a class and are also left "as-is" after data augmentation here:
https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/loader/mri2d_segmentation_dataset.py#L224

Is that correct?

**Comment by charleygros** [comment]
What we called classification: "Is there a lesion in this 2D slice?" --> Yes / No (without need to localise the lesion pixel wise)

Another case is: detection.
Can be done different ways. One way in ivadomed is "bounding box". Cf tumor model.
Another way is 1 pixel, eg at the center of the structure, eg center of the PMJ, eg center of the vertebral body. I need to look at the vertebral detection PR to understand how their GT is constructed.

Sorry for messy answer. Will come back in 3hrs with a more structured one. Hope still makes kinda sense

**Comment by mariehbourget** [comment]
As discussed in meeting, ivadomed assumes that GT label values for MRI segmentation are indeed 0 or 1.
In that case, we will adjust the Microscopy GT values to be 0-1 instead of 0-255 when converting the PNG files to NifTI here https://github.com/ivadomed/ivadomed/blob/7a595ae39e94085837582f7176b8ec59bd9a2bc9/ivadomed/loader/segmentation_pair.py#L294

Those NifTi files are then used for training/testing and evaluation so the values will have the same 0-1 range everywhere.

**Comment by charleygros** [comment]
Cool! Sounds good

My only concern was in [this PR](https://github.com/ivadomed/ivadomed/pull/852), but looking at it: looks like the Class `VertebralSplitting` will create a GT with multiple channels: one per label, only 0s and 1s.
So: all clear
